pkg/listwatch/namespace_denylist.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package listwatch
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// denylistListerWatcher implements cache.ListerWatcher
+// which wraps a cache.ListerWatcher,
+// filtering list results and watch events by denied namespaces.
+type denylistListerWatcher struct {
+	denylist map[string]struct{}
+	next     cache.ListerWatcher
+	logger   log.Logger
+}
+
+// newDenylistListerWatcher creates a cache.ListerWatcher
+// wrapping the given next cache.ListerWatcher
+// filtering lists and watch events by the given namespaces.
+func newDenylistListerWatcher(l log.Logger, namespaces []string, next cache.ListerWatcher) cache.ListerWatcher {
+	denylist := make(map[string]struct{})
+
+	for _, ns := range namespaces {
+		denylist[ns] = struct{}{}
+	}
+
+	return &denylistListerWatcher{
+		denylist: denylist,
+		next:     next,
+		logger:   l,
+	}
+}
+
+func (w *denylistListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	var (
+		l     = metav1.List{}
+		error = level.Error(w.logger)
+		debug = level.Debug(w.logger)
+	)
+
+	list, err := w.next.List(options)
+	if err != nil {
+		error.Log("msg", "error listing", "err", err)
+		return nil, err
+	}
+
+	objs, err := meta.ExtractList(list)
+	if err != nil {
+		error.Log("msg", "error extracting list", "err", err)
+		return nil, err
+	}
+
+	metaObj, err := meta.ListAccessor(list)
+	if err != nil {
+		error.Log("msg", "error getting list accessor", "err", err)
+		return nil, err
+	}
+
+	for _, obj := range objs {
+		acc, err := meta.Accessor(obj)
+		if err != nil {
+			error.Log("msg", "error getting meta accessor accessor", "obj", fmt.Sprintf("%v", obj), "err", err)
+			return nil, err
+		}
+
+		debug = log.With(debug, "name", fmt.Sprintf("%v/%v", acc.GetNamespace(), acc.GetName()))
+
+		if _, denied := w.denylist[acc.GetNamespace()]; denied {
+			debug.Log("msg", "denied")
+			continue
+		}
+
+		debug.Log("msg", "allowed")
+
+		l.Items = append(l.Items, runtime.RawExtension{Object: obj.DeepCopyObject()})
+	}
+
+	l.ListMeta.ResourceVersion = metaObj.GetResourceVersion()
+	return &l, nil
+}
+
```

Comment:
Nit: same as above, adding a function comment would be nice to have.

Reply:
fixed